temporal-sdk/src/main/java/io/temporal/internal/client/ScheduleProtoUtil.java

```diff
@@ -0,0 +1,477 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.internal.client;
+
+import static io.temporal.internal.common.HeaderUtils.toHeaderGrpc;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import io.temporal.api.common.v1.*;
+import io.temporal.api.schedule.v1.*;
+import io.temporal.api.schedule.v1.Schedule;
+import io.temporal.api.schedule.v1.ScheduleAction;
+import io.temporal.api.schedule.v1.ScheduleActionResult;
+import io.temporal.api.schedule.v1.ScheduleInfo;
+import io.temporal.api.schedule.v1.ScheduleSpec;
+import io.temporal.api.schedule.v1.ScheduleState;
+import io.temporal.api.taskqueue.v1.TaskQueue;
+import io.temporal.api.workflow.v1.NewWorkflowExecutionInfo;
+import io.temporal.client.WorkflowOptions;
+import io.temporal.client.schedules.*;
+import io.temporal.common.context.ContextPropagator;
+import io.temporal.common.converter.EncodedValues;
+import io.temporal.internal.client.external.GenericWorkflowClient;
+import io.temporal.internal.common.ProtobufTimeUtils;
+import io.temporal.internal.common.RetryOptionsUtils;
+import io.temporal.internal.common.SearchAttributesUtil;
+import java.time.Instant;
+import java.util.*;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+
+public class ScheduleProtoUtil {
+
+  private final GenericWorkflowClient genericClient;
+  private final ScheduleClientOptions clientOptions;
+
+  public ScheduleProtoUtil(
+      GenericWorkflowClient genericClient, ScheduleClientOptions clientOptions) {
+    this.genericClient = genericClient;
+    this.clientOptions = clientOptions;
+  }
+
+  private io.temporal.common.interceptors.Header extractContextsAndConvertToBytes(
+      List<ContextPropagator> scheduleOptionsContextPropagators) {
+    List<ContextPropagator> scheduleClientContextPropagators =
+        clientOptions.getContextPropagators();
+    if ((scheduleClientContextPropagators.isEmpty() && scheduleOptionsContextPropagators == null)
+        || (scheduleOptionsContextPropagators != null
+            && scheduleOptionsContextPropagators.isEmpty())) {
+      return null;
+    }
+
+    List<ContextPropagator> listToUse =
+        MoreObjects.firstNonNull(
+            scheduleOptionsContextPropagators, scheduleClientContextPropagators);
+    Map<String, Payload> result = new HashMap<>();
+    for (ContextPropagator propagator : listToUse) {
+      result.putAll(propagator.serializeContext(propagator.getCurrentContext()));
+    }
+    return new io.temporal.common.interceptors.Header(result);
+  }
+
+  public ScheduleAction actionToProto(io.temporal.client.schedules.ScheduleAction action) {
```

Comment:
May be clearer to put these helpers on the instances themselves instead of one big utility in a separate package